Fix memory leaks in PortableHostCollection deserialization

makortel opened the pull request:
#### PR description:

Before ROOT allocated the member arrays of `layout_` of the `newObj`, that are then overwritten by `ROOTReadStreamer()` without deallocation. Setting `layout_` as the target as well tells ROOT to not to do those allocations. More details in #40491.

The second commit by @ericcano introduces `Layout::ROOTStreamerCleaner()` to delete the Layout's member arrays allocated by ROOT (more details in https://github.com/cms-sw/cmssw/issues/40491#issuecomment-1398209280).

Resolves #40491

#### PR validation:

Checked with `valgrind` that the amount of memory leaks in `HeterogeneousCore/AlpakaTest/test/reader.py` decreased.

makortel committed:
Fix a memory leak

Before ROOT allocated the member arrays of layout_ of the newObj, that
are then overwritten by ROOTReadStreamer without deallocation. Setting
layout_ as the target as well tells ROOT to not to do those allocations.

Tested with DataFormats/PortableTestObjects, but should apply to
CUDADataFormats/PortableTestObjects as well.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40492/33668

- This PR adds an extra 12KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/PortableTestObjects/src/alpaka/classes_serial_def.xml  modified in PR(s): #40285

cmsbuild commented:
A new Pull Request was created by  @makortel (Matti Kortelainen) for master.

It involves the following packages:

- DataFormats/PortableTestObjects (**heterogeneous**)


@cmsbuild, @makortel, @fwyzard can you please review it and eventually sign? Thanks.
@missirol, @rovere this is something you requested to watch as well.
@perrotta, @dpiparo, @rappoccio you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

makortel commented:
enable gpu

makortel commented:
@cmsbuild, please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0b8d8c/29931/summary.html
**COMMIT**: 44d861b2cbbce0b83a35af4d3d7c8c2722430d17
**CMSSW**: CMSSW_13_0_X_2023-01-11-1100/el8_amd64_gcc11
**Additional Tests**: GPU
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/40492/29931/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



There are some workflows for which there are errors in the baseline:
 140.01 step 3
140.03 step 3
140.104 step 3
140.105 step 3
140.106 step 3
140.107 step 3
140.109 step 3
140.111 step 3
140.112 step 3
140.113 step 3
140.114 step 3
140.115 step 3
140.116 step 3
140.11 step 3
 The results for the comparisons for these workflows could be incomplete 
This means most likely that the IB is having errors in the relvals.The error does NOT come from this pull request

**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 5 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3555538
 - DQMHistoTests: Total failures: 3
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3555513
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 211 log files, 162 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_0_X_2023-01-11-1100+0b8d8c/54957/triggerResults): no differences found


## GPU Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 9 differences found in the comparisons
 - DQMHistoTests: Total files compared: 4
 - DQMHistoTests: Total histograms compared: 19862
 - DQMHistoTests: Total failures: 198
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 19664
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 3 files compared)
 - Checked 12 log files, 9 edm output root files, 4 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_13_0_X_2023-01-11-1100+0b8d8c/54955/triggerResults): no differences found

makortel commented:
Now applying the same fix to `CUDADataFormats/PortableTestObjects/src/classes_def.xml` as well.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40492/33678

- This PR adds an extra 12KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/PortableTestObjects/src/alpaka/classes_serial_def.xml  modified in PR(s): #40285

cmsbuild commented:
Pull request #40492 was updated. @cmsbuild, @makortel, @fwyzard can you please check and sign again.

makortel commented:
@cmsbuild, please test

cmsbuild commented:
-1

**Failed Tests**: RelVals-INPUT 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0b8d8c/29944/summary.html
**COMMIT**: 8192b7f0c36068329561419115d1ebc54c8d7cab
**CMSSW**: CMSSW_13_0_X_2023-01-12-1100/el8_amd64_gcc11
**Additional Tests**: GPU
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/40492/29944/install.sh` to create a dev area with all the needed externals and cmssw changes.


## RelVals-INPUT

The relvals timed out after 4 hours.


## Comparison Summary



There are some workflows for which there are errors in the baseline:
 11634.15 step 3
 The results for the comparisons for these workflows could be incomplete 
This means most likely that the IB is having errors in the relvals.The error does NOT come from this pull request

**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 8 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3555538
 - DQMHistoTests: Total failures: 104
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3555412
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 211 log files, 162 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_0_X_2023-01-12-1100+0b8d8c/54968/triggerResults): no differences found


## GPU Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 9 differences found in the comparisons
 - DQMHistoTests: Total files compared: 4
 - DQMHistoTests: Total histograms compared: 19862
 - DQMHistoTests: Total failures: 197
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 19665
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 3 files compared)
 - Checked 12 log files, 9 edm output root files, 4 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_13_0_X_2023-01-12-1100+0b8d8c/54964/triggerResults): no differences found

makortel commented:
Comparison failures are in 11634.7 (#39803)

ericcano commented:
I did not manage to reproduce the changes. When running a vanilla `CMSSW_13_0_X_2023-01-17-1100` on `lxplus8-gpu`, I get the following summary (test command is: `valgrind --suppressions=Utilities/ReleaseScripts/data/cms-valgrind-memcheck.supp --suppressions=${ROOTSYS}/etc/valgrind-root.supp --suppressions=/usr/share/doc/python38-devel/valgrind-python.supp --leak-check=full  cmsRun HeterogeneousCore/AlpakaTest/test/reader.py`) :

```
==1439048== LEAK SUMMARY:
==1439048==    definitely lost: 498,832 bytes in 227 blocks
==1439048==    indirectly lost: 0 bytes in 0 blocks
==1439048==      possibly lost: 57,312 bytes in 147 blocks
==1439048==    still reachable: 77,509,780 bytes in 136,256 blocks
```

While after adding `layout_` in the list target of the XML configuration for reflex I get:

```
==1452975== LEAK SUMMARY:
==1452975==    definitely lost: 501,100 bytes in 232 blocks
==1452975==    indirectly lost: 16 bytes in 1 blocks
==1452975==      possibly lost: 56,976 bytes in 147 blocks
==1452975==    still reachable: 77,507,384 bytes in 136,248 blocks
```

makortel commented:
I just repeated on `lxplus7-gpu`
```sh
cmsrel CMSSW_13_0_X_2023-01-17-1100
cd CMSSW_13_0_X_2023-01-17-1100/src
cmsenv
git cms-addpkg HeterogeneousCore/AlpakaTest
cmsRun HeterogeneousCore/AlpakaTest/test/writer.py
valgrind --tool=memcheck $(cmsvgsupp) --suppressions=$ROOTSYS/etc/valgrind-root.supp --leak-check=full  cmsRun HeterogeneousCore/AlpakaTest/test/reader.py > valgrind1.txt 2>&1
git cherry-pick 8192b7f0c36068329561419115d1ebc54c8d7cab
git cms-addpkg DataFormats/PortableTestObjects CUDADataFormats/PortableTestObjects
scram b -j 10
valgrind --tool=memcheck $(cmsvgsupp) --suppressions=$ROOTSYS/etc/valgrind-root.supp --leak-check=full  cmsRun HeterogeneousCore/AlpakaTest/test/reader.py > valgrind2.txt 2>&1
```
and get for the IB
```
==1711111== LEAK SUMMARY:
==1711111==    definitely lost: 506,380 bytes in 230 blocks
==1711111==    indirectly lost: 16 bytes in 1 blocks
==1711111==      possibly lost: 50,124 bytes in 145 blocks
==1711111==    still reachable: 77,411,959 bytes in 135,746 blocks
```
and for the IB + commit of this PR
```
==1713970== LEAK SUMMARY:
==1713970==    definitely lost: 252,696 bytes in 111 blocks
==1713970==    indirectly lost: 0 bytes in 0 blocks
==1713970==      possibly lost: 33,144 bytes in 143 blocks
==1713970==    still reachable: 77,472,631 bytes in 136,288 blocks
```

Commit:
Fix a memory leak

This commit adds freeing of memory allocated by ROOT during streaming from disk.
It also declares all members of PortableHostCollections as target (as the streamer fill
them all up).

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40492/33828

- This PR adds an extra 24KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/Portable/interface/PortableHostCollection.h  modified in PR(s): #40285
  - File  DataFormats/PortableTestObjects/src/alpaka/classes_serial_def.xml  modified in PR(s): #40285

cmsbuild commented:
Pull request #40492 was updated. @cmsbuild, @makortel, @fwyzard can you please check and sign again.

makortel commented:
@cmsbuild, please test

makortel commented:
Repeating my test in https://github.com/cms-sw/cmssw/pull/40492#issuecomment-1387385345 with the second commit gives
```
==24530== LEAK SUMMARY:
==24530==    definitely lost: 8 bytes in 1 blocks
==24530==    indirectly lost: 0 bytes in 0 blocks
==24530==      possibly lost: 19,272 bytes in 140 blocks
==24530==    still reachable: 78,045,434 bytes in 171,856 blocks
```

so both leaks reported in https://github.com/cms-sw/cmssw/issues/40491 are now fixed

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0b8d8c/30097/summary.html
**COMMIT**: 0eced58292a00a0a35eade8931e1af428051a383
**CMSSW**: CMSSW_13_0_X_2023-01-20-1100/el8_amd64_gcc11
**Additional Tests**: GPU
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/40492/30097/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3555479
 - DQMHistoTests: Total failures: 0
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3555457
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 211 log files, 162 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_0_X_2023-01-20-1100+0b8d8c/55118/triggerResults): no differences found


## GPU Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 4
 - DQMHistoTests: Total histograms compared: 19862
 - DQMHistoTests: Total failures: 75
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 19787
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 3 files compared)
 - Checked 12 log files, 9 edm output root files, 4 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_13_0_X_2023-01-20-1100+0b8d8c/55115/triggerResults): found differences in 1 / 3 workflows

fwyzard commented on DataFormats/PortableTestObjects/src/alpaka/classes_serial_def.xml:
```diff
@@ -5,7 +5,7 @@
     version="[1-]"
     source="portabletest::TestSoA layout_;"
-    target="buffer_"
+    target="buffer_,layout_,view_"
```
I'm happy if we get rid of memory leaks - but why do we have to specify `layout_` and `view_` in the first place ?
They are declared as transient-only, should ROOT leave them alone ?

ericcano replied:
We are redundant. I would say he semantic here is that the streamer will initialize those variables (which is true), but being transient-only, ROOT will not touch them either. We could decide on both convention (minimum information or exhaustively declaring the member variables we initialize (all, in fact).

makortel replied:
I'm a bit confused, `layout_` is the persisted member, no?

makortel replied:
> * `layout_` is persistent; however we want it to use our user-defined memory, not the one provided by ROOT; is this why we need to list it as a "target" ?

Correct.

> * `view_` is transient, so I would expect ROOT not to allocate any memory for it; why do we need to list it as a "target" ?

The `view_` was indeed not part of the minimal fix suggestion. @pcanal, does including a transient `view_` member in the `target` list have impact on any kind from ROOT side?

ericcano replied:
I would say `buffer_` and `view_` have similar statuses and would prefer have both either listed or not. As they are declared transient in the code, it is most likely redundant (TBC by @pcanal) to list them here, but that gives a complete description of the streamer's behavior.

pcanal replied:
If/since `view_` is also affected by this rule, it make sense to list it as part of the target.  (This let the system knows that if `view_` is directly ask for, it should run the rule.

fwyzard replied:
I'm afraid I'm still confused :-(

Without this change, reflex generates this code for the class dictionary:
```c++
    // Schema evolution read functions
    static void read_PortableHostCollectionlEportabletestcLcLTestSoALayoutlE128cOfalsegRsPgR_0( char* target, TVirtualObject *oldObj )
    {
       //--- Automatically generated variables ---
       struct PortableHostCollectionlEportabletestcLcLTestSoALayoutlE128cOfalsegRsPgR_Onfile {
          portabletest::TestSoA &layout_;
          PortableHostCollectionlEportabletestcLcLTestSoALayoutlE128cOfalsegRsPgR_Onfile(portabletest::TestSoA &onfile_layout_ ): layout_(onfile_layout_) {}
       };
       static Long_t offset_Onfile_PortableHostCollectionlEportabletestcLcLTestSoALayoutlE128cOfalsegRsPgR_layout_ = oldObj->GetClass()->GetDataMemberOffset("layout_");
       char *onfile_add = (char*)oldObj->GetObject();
       PortableHostCollectionlEportabletestcLcLTestSoALayoutlE128cOfalsegRsPgR_Onfile onfile(
          *(portabletest::TestSoA*)(onfile_add+offset_Onfile_PortableHostCollectionlEportabletestcLcLTestSoALayoutlE128cOfalsegRsPgR_layout_) );

       static TClassRef cls("portabletest::TestHostCollection");
       static Long_t offset_buffer_ = cls->GetDataMemberOffset("buffer_");
       optional<PortableHostCollection<portabletest::TestSoALayout<128,false> >::Buffer>& buffer_ = *(optional<PortableHostCollection<portabletest::TestSoALayout<128,false> >::Buffer>*)(target+offset_buffer_);
       portabletest::TestHostCollection* newObj = (portabletest::TestHostCollection*)target;

       // Supress warning message.
       (void)oldObj;
       (void)newObj;

       //--- User's code ---
       portabletest::TestHostCollection::ROOTReadStreamer(newObj, onfile.layout_);
    }
```

With this change, the function becomes
```c++
    // Schema evolution read functions
    static void read_PortableHostCollectionlEportabletestcLcLTestSoALayoutlE128cOfalsegRsPgR_0( char* target, TVirtualObject *oldObj )
    {
       //--- Automatically generated variables ---
       struct PortableHostCollectionlEportabletestcLcLTestSoALayoutlE128cOfalsegRsPgR_Onfile {
          portabletest::TestSoA &layout_;
          PortableHostCollectionlEportabletestcLcLTestSoALayoutlE128cOfalsegRsPgR_Onfile(portabletest::TestSoA &onfile_layout_ ): layout_(onfile_layout_) {}
       };
       static Long_t offset_Onfile_PortableHostCollectionlEportabletestcLcLTestSoALayoutlE128cOfalsegRsPgR_layout_ = oldObj->GetClass()->GetDataMemberOffset("layout_");
       char *onfile_add = (char*)oldObj->GetObject();
       PortableHostCollectionlEportabletestcLcLTestSoALayoutlE128cOfalsegRsPgR_Onfile onfile(
          *(portabletest::TestSoA*)(onfile_add+offset_Onfile_PortableHostCollectionlEportabletestcLcLTestSoALayoutlE128cOfalsegRsPgR_layout_) );

       static TClassRef cls("portabletest::TestHostCollection");
       static Long_t offset_buffer_ = cls->GetDataMemberOffset("buffer_");
       optional<PortableHostCollection<portabletest::TestSoALayout<128,false> >::Buffer>& buffer_ = *(optional<PortableHostCollection<portabletest::TestSoALayout<128,false> >::Buffer>*)(target+offset_buffer_);
       static Long_t offset_layout_ = cls->GetDataMemberOffset("layout_");
       PortableHostCollection<portabletest::TestSoALayout<128,false> >::Layout& layout_ = *(PortableHostCollection<portabletest::TestSoALayout<128,false> >::Layout*)(target+offset_layout_);
       static Long_t offset_view_ = cls->GetDataMemberOffset("view_");
       PortableHostCollection<portabletest::TestSoALayout<128,false> >::View& view_ = *(PortableHostCollection<portabletest::TestSoALayout<128,false> >::View*)(target+offset_view_);
       portabletest::TestHostCollection* newObj = (portabletest::TestHostCollection*)target;

       // Supress warning message.
       (void)oldObj;
       (void)newObj;

       //--- User's code ---
       portabletest::TestHostCollection::ROOTReadStreamer(newObj, onfile.layout_);
    }
```

The additional code is
```c++
      static Long_t offset_layout_ = cls->GetDataMemberOffset("layout_");
      PortableHostCollection<portabletest::TestSoALayout<128,false> >::Layout& layout_ = *(PortableHostCollection<portabletest::TestSoALayout<128,false> >::Layout*)(target+offset_layout_);
      static Long_t offset_view_ = cls->GetDataMemberOffset("view_");
      PortableHostCollection<portabletest::TestSoALayout<128,false> >::View& view_ = *(PortableHostCollection<portabletest::TestSoALayout<128,false> >::View*)(target+offset_view_);
```
which simply assigns pointers into the red buffer to references to `layout_` and `view_`.

But I don't see anything anywhere about not allocating memory for them, or deleting them.
In fact, these references are not used at all.

Why should this make any difference ?

pcanal replied:
The difference is not in the generated code.  (The generator does not parse the user `code` for the rule and thus set a series of pointer `just in case` for old the dependencies).

The difference comes from how the `TStreamerInfo` infrastructure is handling the dependencies.

With a struct like:
```
struct Example {
   InnerClass x;
   Transient z;  //!
};
```
A rule listing:
```
source="InnerClass x" target="z"
```
'just' say that `x read from the file` is the input for z.  Consequently it *also* need to read `x` (a second time) to update the in memory x.

A rule listing:
```
source="InnerClass x" target="x,z"
```
says that rule will update `x and z` from the input and there is no need to read it a second time.

Another example:

```
struct Example2
{
     float x;
     float y;
     float z;
};
```
If on file with have `x==2` and `y==5`
with the rules
```
source="float x" target="x" code="{ x = onfile.x * 2; }"
source="float x" target="z" code="{ z =  onfile.x; };
```
after reading we have `x==4` and `z==2`

In other word the `onfile.MemberName` and `MemberName` are independent and unless it was indicated that `MemberName` will be explicitly set by a rule (i.e. is a target) then we have to read it a second time if it is use as input of a rule (because we need to give the rule the not updated version).

fwyzard replied:
I see, thanks for the explanation.

makortel commented:
@fwyzard Do you have any further concerns, or could we proceed with this PR? I believe it would be useful to have the fixes in pre4.

makortel commented:
@cmsbuild, please test

Let's refresh

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0b8d8c/30389/summary.html
**COMMIT**: 0eced58292a00a0a35eade8931e1af428051a383
**CMSSW**: CMSSW_13_0_X_2023-02-03-1100/el8_amd64_gcc11
**Additional Tests**: GPU
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/40492/30389/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially removed 10 lines from the logs
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3555495
 - DQMHistoTests: Total failures: 0
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3555473
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 213 log files, 164 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_0_X_2023-02-03-1100+0b8d8c/55346/triggerResults): no differences found


## GPU Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 4
 - DQMHistoTests: Total histograms compared: 19862
 - DQMHistoTests: Total failures: 67
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 19795
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 3 files compared)
 - Checked 12 log files, 9 edm output root files, 4 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_13_0_X_2023-02-03-1100+0b8d8c/55342/triggerResults): found differences in 2 / 3 workflows

fwyzard commented:
> @fwyzard Do you have any further concerns, or could we proceed with this PR? I believe it would be useful to have the fixes in pre4.

I'll have another look on Monday.

fwyzard commented:
please test

fwyzard commented:
+heterogeneous

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

rappoccio commented:
+1
